api/eth_types.go

```diff
@@ -156,22 +185,71 @@ func (c *EthCall) UnmarshalJSON(b []byte) error {
 }
 
 type EthTxReceipt struct {
-	TransactionHash  EthHash    `json:"transactionHash"`
-	TransactionIndex EthInt     `json:"transacionIndex"`
-	BlockHash        EthHash    `json:"blockHash"`
-	BlockNumber      EthHash    `json:"blockNumber"`
-	From             EthAddress `json:"from"`
-	To               EthAddress `json:"to"`
+	TransactionHash  EthHash     `json:"transactionHash"`
+	TransactionIndex EthInt      `json:"transacionIndex"`
+	BlockHash        EthHash     `json:"blockHash"`
+	BlockNumber      EthInt      `json:"blockNumber"`
+	From             EthAddress  `json:"from"`
+	To               *EthAddress `json:"to"`
 	// Logs
 	// LogsBloom
 	StateRoot         EthHash     `json:"root"`
 	Status            EthInt      `json:"status"`
 	ContractAddress   *EthAddress `json:"contractAddress"`
-	CumulativeGasUsed EthBigInt   `json:"cumulativeGasUsed"`
-	GasUsed           EthBigInt   `json:"gasUsed"`
+	CumulativeGasUsed EthInt      `json:"cumulativeGasUsed"`
+	GasUsed           EthInt      `json:"gasUsed"`
 	EffectiveGasPrice EthBigInt   `json:"effectiveGasPrice"`
 }
 
+func NewEthTxReceipt(tx EthTx, lookup *MsgLookup, replay *InvocResult) (EthTxReceipt, error) {
+	receipt := EthTxReceipt{
+		TransactionHash:  tx.Hash,
+		TransactionIndex: tx.TransactionIndex,
+		BlockHash:        tx.BlockHash,
+		BlockNumber:      tx.BlockNumber,
+		From:             tx.From,
+		To:               tx.To,
+		StateRoot:        EmptyEthHash,
+	}
+
+	contractAddr, err := CheckContractCreation(lookup)
+	if err == nil {
+		receipt.To = nil
+		receipt.ContractAddress = contractAddr
+	}
+
+	if lookup.Receipt.ExitCode.IsSuccess() {
+		receipt.Status = 1
+	}
+	if lookup.Receipt.ExitCode.IsError() {
+		receipt.Status = 0
+	}
+
+	receipt.GasUsed = EthInt(lookup.Receipt.GasUsed)
+
+	// TODO: handle CumulativeGasUsed
+	receipt.CumulativeGasUsed = EmptyEthInt
+
+	effectiveGasPrice := big.Div(replay.GasCost.TotalCost, big.NewInt(lookup.Receipt.GasUsed))
+	receipt.EffectiveGasPrice = EthBigInt(effectiveGasPrice)
+	return receipt, nil
+}
+
+func CheckContractCreation(lookup *MsgLookup) (*EthAddress, error) {
+	if lookup.Receipt.ExitCode.IsError() {
+		return nil, xerrors.Errorf("message execution was not successful")
+	}
+	var result init8.ExecReturn
+	ret := bytes.NewReader(lookup.Receipt.Return)
+	if err := result.UnmarshalCBOR(ret); err == nil {
+		contractAddr, err := EthAddressFromFilecoinIDAddress(result.IDAddress)
```

Comment:
nit (can safe for a future PR): should probably return a specific error message instead of letting it fall through to the "not a contract creation tx" error.